### PR TITLE
Support OpenSearch Serverless

### PR DESCRIPTION
**Pull request by kukushking**
### Feature or Bugfix
- Feature

### Detail
- Add OpenSearch Serverless stack & corresponding feature flag
- Update client request signature ("aoss"/"es") and create service SSM parameter
- Update domain / collection references
- Add ECS task role to OpenSearch Serverless principals
- Update CDK library to latest (2.61.1)
- Update Pipeline CodeBuild image to `AMAZON_LINUX_2_4` with latest node version to use latest CDK and remove unnecessary packages
- Pass VPC endpoints security group to OpenSearch Serverless stack
- Update template_cdk.json
- Add monitoring

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: deploy/stacks/monitoring.py
Comment (dlpzx) [inline deploy/stacks/monitoring.py:140]: What about when there is no opensearch_domain? no alarm?

File: deploy/stacks/monitoring.py
Comment (dlpzx) [inline deploy/stacks/monitoring.py:114]: this is a typo right? opensearch

Comment (kukushking) [reply deploy/stacks/monitoring.py:114]: Yes, thanks for the catch!

File: deploy/stacks/monitoring.py
Comment (kukushking) [reply deploy/stacks/monitoring.py:140]: If there is no `opensearch_domain`, there should be `opensearch_serverless_collection_id` that has separate monitoring, including alarms on availability of the collection (collections are domains in serverless). The reason I separated the two is because serverless provides a different set of metrics. 

I can either put serverless monitoring into else clause here or raise an error if none of these two parameters are provided - whichever works best for you?

File: template_cdk.json
Comment (kukushking) [reply template_cdk.json:35]: We did discuss it with @dlpzx and decided to keep it backwards-compatible and rather opt-in using a feature flag. Besides, at the time of the PR publication AOSS wasn't even GA yet. Happy to discuss this further though.

File: deploy/stacks/backend_stack.py
Comment (kukushking) [reply deploy/stacks/backend_stack.py:274]: Thanks!

File: deploy/stacks/opensearch.py
Comment (kukushking) [reply deploy/stacks/opensearch.py:27]: Not in server-full stack, no.

File: deploy/stacks/monitoring.py
Comment (kukushking) [reply deploy/stacks/monitoring.py:447]: Agreed, thanks.

File: template_cdk.json
Comment (nikpodsh) [reply template_cdk.json:35]: Makes sense. Let's keep it as it is. Thanks for the clarification!

File: deploy/stacks/backend_stack.py
Comment (kukushking) [reply deploy/stacks/backend_stack.py:293]: I've implemented your suggestion with the only exception which is `create_opensearch_serverless_stack` is still called by the backend stack based on the passed feature flag. We do follow the same approach for other feature flags as well. The main reason is that I want backend stack to be self-contained to be able to synth/deploy it w/o having to rely on the stage having to call either of the opensearch methods.  Let me know if that works!

File: deploy/stacks/opensearch.py
Comment (kukushking) [reply deploy/stacks/opensearch.py]: Removed

File: deploy/stacks/monitoring.py
Comment (kukushking) [reply deploy/stacks/monitoring.py]: Updated

File: deploy/stacks/backend_stack.py
Comment (nikpodsh) [reply deploy/stacks/backend_stack.py:293]: Sure, looks good! Thank you for addressing it.

File: deploy/stacks/pipeline.py
Comment (nikpodsh) [inline deploy/stacks/pipeline.py:214]: Looking into https://github.com/aws/aws-codebuild-docker-images/blob/master/ubuntu/standard/6.0/Dockerfile I couldn't find `shadow-utils` (should be `libvshadow-utils` in Ubuntu) and something to related to `postgre` (since we are deleting `postgresql-devel`). Are they shipped with `STANDARD_6_0`?
I am not sure if we use these tools at all at the moment , but wanted to double check with you that we removed them intentionally and there is no need for something like `apt-get install -y libvshadow-utils` 🙂

File: deploy/stacks/backend_stack.py
Comment (nikpodsh) [inline deploy/stacks/backend_stack.py:339]: Super minor: looks like `set_aoss_alarms` is called only from this place and we don't need `set_aoss_alarms` to have a parameter `cw_alarm_action` since we can use it directly from `monitoring_stack`

File: deploy/stacks/pipeline.py
Comment (kukushking) [reply deploy/stacks/pipeline.py:214]: Tbh I wasn't sure if those are needed as well. This specific CodeBuild is only responsible for running `cdk synth` so they seemed unnecessary. It's worth to double check with @dlpzx though.

On a related note, since I've updated this one, might as well upgrade other CBs to `STANDARD_6_0` but I will wait for the confirmation first.

File: deploy/stacks/backend_stack.py
Comment (kukushking) [reply deploy/stacks/backend_stack.py]: Thanks, resolved

**Review by nikpodsh — APPROVED** [review]
Since [this](https://github.com/awslabs/aws-dataall/pull/292#discussion_r1102877558) seems to work fine, I think we can go without it:)
Approved.  Great job! 👍

**Comment by kukushking** [comment]
Thanks for the review @nikpodsh!

File: deploy/stacks/pipeline.py
Comment (dlpzx) [reply deploy/stacks/pipeline.py:214]: The reason behind installing node directly in CodeBuild is that node16 is needed for CDK2 and it was not available at runtime in `AMAZON_LINUX_2_3` images. With `AMAZON_LINUX_2_4` node16 is installed at runtime so we can remove those lines. At first sight I would prefer to use `AMAZON_LINUX_2_4` instead of Ubuntu to avoid changes that could cause errors (plus documentation changes), but maybe I am missing something.

I am also testing the libraries that we are installing ('yum -y install shadow-utils wget && yum -y install openssl-devel bzip2-devel libffi-devel postgresql-devel') caused by the networking changes imposed in #307.

**Review by dlpzx — COMMENTED** [review]
Before merging we can include what we discussed yesterday around existing customers that want to take advantage of this feature.

How do we migrate the information from OpenSearch non-serverless to serverless. At least we need to provide some info

File: deploy/stacks/pipeline.py
Comment (kukushking) [reply deploy/stacks/pipeline.py:214]: Thanks @dlpzx, makes sense. Let me give `AMAZON_LINUX_2_4` a try. I'll also keep the libraries to reduce blast radius.

Comment (kukushking) [reply deploy/stacks/pipeline.py:214]: Upgraded all CBs to `AMAZON_LINUX_2_4` & performed a test deployment - all good.

Comment (kukushking) [reply deploy/stacks/pipeline.py:214]: @dlpzx all of the packages are already in 4.0 image so we're good removing them. Removed & ran through the deployment pipeline to verify 👍

**Comment by dlpzx** [comment]
Thank you for cleaning up the unnecessary packages from the CICD pipeline! The testing in v1m5m0 branch is ready, so you can go ahead and merge your changes. Great contribution @kukushking 🚀

**Pull request merged**
